sotodlib/site_pipeline/update_smurf_caldbs.py

```diff
@@ -109,15 +123,252 @@ def smurf_detset_info(config:str | dict,
+        if substr in f:
+            fpath = os.path.join(smurf_dir, f)
+            break
+    else:
+        raise FileNotFoundError("Could not find npy file")
+    res = np.load(fpath, allow_pickle=True).item()
+    return res
+
+
+def get_cal_resset(ctx: core.Context, obs_id):
+    """Returns calibration ResultSet for a given ObsId"""
+    am = ctx.get_obs(obs_id, samples=(0, 1), ignore_missing=True)
+
+    cals = [CalInfo(rid) for rid in am.det_info.readout_id]
+
+    iv_obsids = get_cal_obsids(ctx, obs_id, 'iv')
+
+    rtm_bit_to_volt = None
+    ivas = {dset: None for dset in iv_obsids}
+    for dset, oid in iv_obsids.items():
+        if oid is not None:
+            ivas[dset] = _load_smurf_npy(oid, 'iv')
+            if rtm_bit_to_volt is None:
+                rtm_bit_to_volt = ivas[dset]['meta']['rtm_bit_to_volt']
+
+    bias_step_obsids = get_cal_obsids(ctx, obs_id, 'bias_steps')
+    bsas = {dset: None for dset in bias_step_obsids}
+    for dset, oid in bias_step_obsids.items():
+        if oid is not None:
+            bsas[dset] = _load_smurf_npy(oid, 'bias_step_analysis')
+            if rtm_bit_to_volt is None:
+                rtm_bit_to_volt = bsas[dset]['meta']['rtm_bit_to_volt']
+
+    rtm_bit_to_volt = DEFAULT_RTM_BIT_TO_VOLT
+
+    # Add IV info
+    for i, cal in enumerate(cals):
+        band = am.det_info.smurf.band[i]
+        chan = am.det_info.smurf.channel[i]
+        detset = am.det_info.detset[i]
+        iva = ivas[detset]
+
+        if iva is None: # No IV analysis for this detset
+            continue
+
+        ridx = np.where(
+            (iva['bands'] == band) & (iva['channels'] == chan)
+        )[0]
+        if not ridx: # Channel doesn't exist in IV analysis
+            continue
+
+        ridx = ridx[0]
+        cal.bg = iva['bgmap'][ridx]
+        cal.polarity = iva['polarity'][ridx]
+        cal.r_n = iva['R_n'][ridx]
+        cal.p_sat = iva['p_sat'][ridx]
+    
+    obs_biases = dict(zip(am.bias_lines.vals, am.biases[:, 0] * 2*rtm_bit_to_volt))
+    for i, cal in enumerate(cals):
+        band = am.det_info.smurf.band[i]
+        chan = am.det_info.smurf.channel[i]
+        detset = am.det_info.detset[i]
+        stream_id = am.det_info.stream_id[i]
+        bg = cal.bg
+        bsa = bsas[detset]
+
+        if bsa is None or bg == -1:
+            continue
+
+        bl_label = f'{stream_id}_b{bg:0>2}'
+        # If observation bias differs from bias-steps by more than 0.1 V,
+        # don't include bias step calibration info
+        if np.abs(obs_biases[bl_label] - bsa['Vbias'][bg]) > 0.1:
+            continue
+
+        ridx = np.where(
+            (bsa['bands'] == band) & (bsa['channels'] == chan)
+        )[0]
+        if not ridx: # Channel doesn't exist in bias step analysis
+            continue
+
+        ridx = ridx[0]
+        cal.r_tes = bsa['R0'][ridx]
+        cal.r_frac = bsa['Rfrac'][ridx]
+        cal.p_bias = bsa['Pj'][ridx]
+        cal.s_i = bsa['Si'][ridx]
```

Comment:
Can we add a field that is `cal.dac_to_pW = (9e6/2/np.pi) / (bsa['Si'][ridx]*bsa['polarity'][ridx])` I feel like for folks not knowledgeable of the detectors it would be good to just multiply by one obviously named parameter to get into pW units.

Reply:
Sure, this should be `phase_to_pW` though, right?

Reply:
Ya that sounds like a reasonable name.